### ISSUE TITLE
Segmentation fault

### ISSUE THREAD

**Issue by diego-rt** [nanoporetech/dorado#514]
Hi,

I'm getting a bunch of segmentation faults on my data. Sometimes they go away when changing the GPU model, but not always, so I suspect an issue with the batch size or memory management. The data is ultra long data.

This the command:

```
dorado duplex -r sup@v4.3.0,5mCG_5hmCG chunk_pod5/ > R020.basecalls.bam
```

This is the output:

```
[2023-12-10 14:19:17.913] [info] > No duplex pairs file provided, pairing will be performed automatically
[2023-12-10 14:19:26.298] [info]  - set batch size for cuda:0 to 1472
[2023-12-10 14:19:29.205] [info]  - set batch size for cuda:0 to 2304
[2023-12-10 14:19:29.205] [info] > Starting Stereo Duplex pipeline
[2023-12-10 14:19:29.225] [info] > Reading read channel info
[2023-12-10 14:19:29.373] [info] > Processed read channel info
/scratch/1_processData/2_Nanopore-Ultralong/1_basecalling/19/4d2f8687bab6db5b33030a9ae636cd/.command.sh: line 7: 36960 Segmentation fault      dorado duplex -r sup@v4.3.0,5mCG_5hmCG chunk_pod5/ > R020.basecalls.bam
```

The GPUs tested are A100, TU102 (RTX 6000) and V100.

**Comment by tijyojwad** [comment]
Hi @diego-rt - thank you for reporting this!

Is this always encountered with mod base calling enabled? If you run without mod bees does it go away?

**Comment by diego-rt** [comment]
Hey thanks for the quick reply!

You are totally right, it goes away after removing the modification calling. I managed to get most of my jobs to finish by moving them to the RTX6000 or V100 rather than the A100, but there are two which have failed regardless of the GPU and I only managed to get them to finish right now by removing the modification calling. 

Is there any parameter that you suspect I can tweak to get them to finish with modifications on? batch size maybe?

**Comment by diego-rt** [comment]
Seems like chunk size `-c 5000` got basecalling with modification to bypass the segfault, but not reducing the batch size on an A100. Is it a good idea to modify this setting?

**Comment by tijyojwad** [comment]
Adjusting the chunk size is not a good idea since it will affect the accuracy of the output. We will look into the seg fault issue with duplex mod basecalling.

Can you post `stderr` logs just before the crash?

**Comment by diego-rt** [comment]
Sure, here it goes. 

Thanks for looking into it!

[dorado.log](https://github.com/nanoporetech/dorado/files/13630057/dorado.log)

**Comment by pustoshilov-d** [comment]
Hi! I have faced the same error for simplex basecalling

Run some tests with 18gb dataset 
2x Tesla V100S
NVIDIA-SMI 525.60.11    Driver Version: 525.60.11    CUDA Version: 12.0

|                 |       |
|-----------------|-------|
| sup@v4.3_A_C_CG | failed  |
| sup@v4.3_CG_C_A   | failed  |
| sup@v4.3_CG_C   | failed  |
| sup@v4.3_A_C    | ended |
| sup@v4.3_A_CG   | ended |
| sup@v4.3_A   | ended |
| sup@v4.3_C   | ended |
| sup@v4.3_CG_A| ended |
| sup@v4.3_CG  | ended |
| sup@v4.3  | ended |

It looks like that problem is with using mods CG and C at the same time in my case. 
@tijyojwad Should I create a new issue?

stderr atteached

[ONT.20231110.dorado_sup@v4.3_CG_C.minimap2.txt](https://github.com/nanoporetech/dorado/files/13646116/ONT.20231110.dorado_sup%40v4.3_CG_C.minimap2.txt)
[ONT.20231110.dorado_sup@v4.3_A_C_CG.minimap2.txt](https://github.com/nanoporetech/dorado/files/13646118/ONT.20231110.dorado_sup%40v4.3_A_C_CG.minimap2.txt)

**Comment by ulbivin** [comment]
Any update on this? I have the same issue on A100 cards.

**Comment by tijyojwad** [comment]
Hi @diego-rt - we just released v0.5.1 (https://github.com/nanoporetech/dorado#installation) - can you give that a try?

**Comment by diego-rt** [comment]
Hey @tijyojwad 

I gave it a try but unfortunately no luck. I narrowed it down to two specific files though. The rest of the crashes I managed to solve by switching between GPU models (unfortunately I have no easy way of tracking down which files those were). 

Here is an example log:

```
[diego.terrones@clip-g4-6 seg_fault_pod5_R43]$ singularity exec --cleanenv --nv docker://docker.artifactory.imp.ac.at/tanakalab/docker-dorado:0.5.1 dorado duplex -v -r sup@v4.3.0,5mCG_5hmCG filtered.channel_2159.pod5 > TEST.basecalls.bam
INFO:    Using cached SIF image
[2023-12-23 17:43:19.483] [info] > No duplex pairs file provided, pairing will be performed automatically
[2023-12-23 17:43:19.537] [debug] > Reads to process: 636
[2023-12-23 17:43:19.566] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0 with httplib
[2023-12-23 17:43:22.293] [info]  - downloading dna_r10.4.1_e8.2_5khz_stereo@v1.2 with httplib
[2023-12-23 17:43:22.799] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v4.3.0_5mCG_5hmCG@v1 with httplib
[2023-12-23 17:43:26.067] [debug] cuda:0 memory available: 41.26GB
[2023-12-23 17:43:26.067] [debug] Auto batchsize cuda:0: memory limit 36.14GB
[2023-12-23 17:43:26.067] [debug] Auto batchsize cuda:0: testing up to 1472 in steps of 64
[2023-12-23 17:43:26.605] [debug] Auto batchsize cuda:0: 64, time per chunk 1.038256 ms
[2023-12-23 17:43:26.749] [debug] Auto batchsize cuda:0: 128, time per chunk 0.556544 ms
[2023-12-23 17:43:26.875] [debug] Auto batchsize cuda:0: 192, time per chunk 0.324459 ms
[2023-12-23 17:43:27.005] [debug] Auto batchsize cuda:0: 256, time per chunk 0.253268 ms
[2023-12-23 17:43:27.141] [debug] Auto batchsize cuda:0: 320, time per chunk 0.212099 ms
[2023-12-23 17:43:27.294] [debug] Auto batchsize cuda:0: 384, time per chunk 0.198149 ms
[2023-12-23 17:43:27.436] [debug] Auto batchsize cuda:0: 448, time per chunk 0.157433 ms
[2023-12-23 17:43:27.576] [debug] Auto batchsize cuda:0: 512, time per chunk 0.136786 ms
[2023-12-23 17:43:27.714] [debug] Auto batchsize cuda:0: 576, time per chunk 0.119847 ms
[2023-12-23 17:43:27.856] [debug] Auto batchsize cuda:0: 640, time per chunk 0.110835 ms
[2023-12-23 17:43:28.001] [debug] Auto batchsize cuda:0: 704, time per chunk 0.102330 ms
[2023-12-23 17:43:28.147] [debug] Auto batchsize cuda:0: 768, time per chunk 0.094959 ms
[2023-12-23 17:43:28.292] [debug] Auto batchsize cuda:0: 832, time per chunk 0.087392 ms
[2023-12-23 17:43:28.470] [debug] Auto batchsize cuda:0: 896, time per chunk 0.098718 ms
[2023-12-23 17:43:28.651] [debug] Auto batchsize cuda:0: 960, time per chunk 0.094169 ms
[2023-12-23 17:43:28.835] [debug] Auto batchsize cuda:0: 1024, time per chunk 0.089848 ms
[2023-12-23 17:43:29.023] [debug] Auto batchsize cuda:0: 1088, time per chunk 0.086258 ms
[2023-12-23 17:43:29.218] [debug] Auto batchsize cuda:0: 1152, time per chunk 0.084241 ms
[2023-12-23 17:43:29.427] [debug] Auto batchsize cuda:0: 1216, time per chunk 0.085997 ms
[2023-12-23 17:43:29.641] [debug] Auto batchsize cuda:0: 1280, time per chunk 0.083555 ms
[2023-12-23 17:43:30.006] [debug] Auto batchsize cuda:0: 1344, time per chunk 0.083611 ms
[2023-12-23 17:43:30.242] [debug] Auto batchsize cuda:0: 1408, time per chunk 0.082977 ms
[2023-12-23 17:43:30.488] [debug] Auto batchsize cuda:0: 1472, time per chunk 0.082820 ms
[2023-12-23 17:43:30.488] [debug] Device cuda:0 Model memory 25.11GB
[2023-12-23 17:43:30.488] [debug] Device cuda:0 Decode memory 10.38GB
[2023-12-23 17:43:30.533] [info]  - set batch size for cuda:0 to 1472
[2023-12-23 17:43:30.533] [debug] - adjusted chunk size to match model stride: 10000 -> 9996
[2023-12-23 17:43:30.732] [debug] cuda:0 memory available: 41.50GB
[2023-12-23 17:43:30.732] [debug] Auto batchsize cuda:0: memory limit 19.75GB
[2023-12-23 17:43:30.732] [debug] Auto batchsize cuda:0: testing up to 2432 in steps of 64
[2023-12-23 17:43:30.760] [debug] Auto batchsize cuda:0: 64, time per chunk 0.191664 ms
[2023-12-23 17:43:30.786] [debug] Auto batchsize cuda:0: 128, time per chunk 0.098800 ms
[2023-12-23 17:43:30.812] [debug] Auto batchsize cuda:0: 192, time per chunk 0.067467 ms
[2023-12-23 17:43:30.839] [debug] Auto batchsize cuda:0: 256, time per chunk 0.052256 ms
[2023-12-23 17:43:30.867] [debug] Auto batchsize cuda:0: 320, time per chunk 0.043498 ms
[2023-12-23 17:43:30.896] [debug] Auto batchsize cuda:0: 384, time per chunk 0.037547 ms
[2023-12-23 17:43:30.926] [debug] Auto batchsize cuda:0: 448, time per chunk 0.033513 ms
[2023-12-23 17:43:30.957] [debug] Auto batchsize cuda:0: 512, time per chunk 0.030084 ms
[2023-12-23 17:43:30.990] [debug] Auto batchsize cuda:0: 576, time per chunk 0.027406 ms
[2023-12-23 17:43:31.031] [debug] Auto batchsize cuda:0: 640, time per chunk 0.032034 ms
[2023-12-23 17:43:31.073] [debug] Auto batchsize cuda:0: 704, time per chunk 0.029564 ms
[2023-12-23 17:43:31.116] [debug] Auto batchsize cuda:0: 768, time per chunk 0.027852 ms
[2023-12-23 17:43:31.159] [debug] Auto batchsize cuda:0: 832, time per chunk 0.026097 ms
[2023-12-23 17:43:31.204] [debug] Auto batchsize cuda:0: 896, time per chunk 0.024639 ms
[2023-12-23 17:43:31.248] [debug] Auto batchsize cuda:0: 960, time per chunk 0.022948 ms
[2023-12-23 17:43:31.293] [debug] Auto batchsize cuda:0: 1024, time per chunk 0.021823 ms
[2023-12-23 17:43:31.339] [debug] Auto batchsize cuda:0: 1088, time per chunk 0.020949 ms
[2023-12-23 17:43:31.386] [debug] Auto batchsize cuda:0: 1152, time per chunk 0.020273 ms
[2023-12-23 17:43:31.444] [debug] Auto batchsize cuda:0: 1216, time per chunk 0.023699 ms
[2023-12-23 17:43:31.502] [debug] Auto batchsize cuda:0: 1280, time per chunk 0.022782 ms
[2023-12-23 17:43:31.561] [debug] Auto batchsize cuda:0: 1344, time per chunk 0.021982 ms
[2023-12-23 17:43:31.621] [debug] Auto batchsize cuda:0: 1408, time per chunk 0.021237 ms
[2023-12-23 17:43:31.682] [debug] Auto batchsize cuda:0: 1472, time per chunk 0.020521 ms
[2023-12-23 17:43:31.744] [debug] Auto batchsize cuda:0: 1536, time per chunk 0.020048 ms
[2023-12-23 17:43:31.807] [debug] Auto batchsize cuda:0: 1600, time per chunk 0.019665 ms
[2023-12-23 17:43:31.871] [debug] Auto batchsize cuda:0: 1664, time per chunk 0.019190 ms
[2023-12-23 17:43:31.936] [debug] Auto batchsize cuda:0: 1728, time per chunk 0.018772 ms
[2023-12-23 17:43:32.011] [debug] Auto batchsize cuda:0: 1792, time per chunk 0.020738 ms
[2023-12-23 17:43:32.086] [debug] Auto batchsize cuda:0: 1856, time per chunk 0.020307 ms
[2023-12-23 17:43:32.163] [debug] Auto batchsize cuda:0: 1920, time per chunk 0.019917 ms
[2023-12-23 17:43:32.241] [debug] Auto batchsize cuda:0: 1984, time per chunk 0.019559 ms
[2023-12-23 17:43:32.320] [debug] Auto batchsize cuda:0: 2048, time per chunk 0.019216 ms
[2023-12-23 17:43:32.400] [debug] Auto batchsize cuda:0: 2112, time per chunk 0.018897 ms
[2023-12-23 17:43:32.481] [debug] Auto batchsize cuda:0: 2176, time per chunk 0.018630 ms
[2023-12-23 17:43:32.563] [debug] Auto batchsize cuda:0: 2240, time per chunk 0.018329 ms
[2023-12-23 17:43:32.647] [debug] Auto batchsize cuda:0: 2304, time per chunk 0.018113 ms
[2023-12-23 17:43:32.740] [debug] Auto batchsize cuda:0: 2368, time per chunk 0.019605 ms
[2023-12-23 17:43:32.835] [debug] Auto batchsize cuda:0: 2432, time per chunk 0.019358 ms
[2023-12-23 17:43:32.835] [debug] Device cuda:0 Model memory 12.98GB
[2023-12-23 17:43:32.835] [debug] Device cuda:0 Decode memory 5.35GB
[2023-12-23 17:43:33.435] [info]  - set batch size for cuda:0 to 2304
[2023-12-23 17:43:33.435] [info] > Starting Stereo Duplex pipeline
[2023-12-23 17:43:33.439] [debug] Using dorado duplex channel count of 10
[2023-12-23 17:43:33.452] [info] > Reading read channel info
[2023-12-23 17:43:33.454] [info] > Processed read channel info
[2023-12-23 17:43:33.455] [debug] Sort channel 2159
[2023-12-23 17:43:33.455] [debug] Sorted channel 2159
Segmentation fault████████████▋] 99% [00m:03s<00m:00s] 
```

The GPUs tested were A100, TU102 (RTX 6000) and V100. Neither of them worked.

Happy to share the two problematic pod5s if you want. Just let me know how I should get them to you.

**Comment by tijyojwad** [comment]
Pod5s would be fantastic! Thanks so much!

**Comment by diego-rt** [comment]
Here they go! [POD5](https://biocenterat-my.sharepoint.com/:u:/g/personal/diego_terrones_imp_ac_at/EXwLaBokQ4FIjV27k-jv8UcBB39_jamdCpbYxVKef0dfcA?e=KYRXiX)

Thanks a lot!

**Comment by tijyojwad** [comment]
Thank you so much! We'll take a look at these POD5s

**Comment by vellamike** [comment]
Hi @diego-rt just a quick note to let you know that I've successfully reproduced your issue. I'm now attempting to establish the cause.

**Comment by ulbivin** [comment]
Hi @tijyojwad and @vellamike 
Has this issue been resolved yet?

**Comment by tijyojwad** [comment]
Hi @diego-rt and @ulbivin - yes sorry, we forgot to update this issue! A fix for this was released as of dorado v0.5.2 and verified with the dataset shared earlier.

Closing the issue now, but if you encounter it again please feel free to re-open!
